File: src/fragment/MapFragment.ts
import { HTMLText, Text } from "pixi.js";
import Fragment, { FragmentType, register } from "./Fragment";
import { StructEndecBuilder, PrimitiveEndecs, KtList, mapEndecOf, KtMap } from 'KEndec';


export default class MapFragment extends Fragment {
    map: ReadonlyMap<Fragment, Fragment>;

    constructor(map: ReadonlyMap<Fragment, Fragment>) {
        super();

        this.map = map;
    }

    override asFormattedText(): HTMLText {
        let out = "{";

        const entries = this.map.entries();
        const iterator = entries[Symbol.iterator]();

        let result = iterator.next();
        while (!result.done) {
            const [key, value] = result.value;
            out += key.asFormattedText().text + ": " + value.asFormattedText().text;

            result = iterator.next();
            if (!result.done) {
                out += ", ";
            }
        }

        out += "}";

        return new HTMLText({
            text: `<span style="color: #${this.type().color.toString(16)}">${out}</span>`,
        });
    }

    override toString(): string {
        let out = "{";

        const entries = this.map.entries();
        const iterator = entries[Symbol.iterator]();

        let result = iterator.next();
        while (!result.done) {
            const [key, value] = result.value;
            out += key.toString() + ": " + value.toString();

            result = iterator.next();
            if (!result.done) {
                out += ", ";
            }
        }

        out += "}";

        return out;
    }

    override type(): FragmentType<MapFragment> {
        return MAP;
    }
}

const MAP = register("map", 0xffffff, 
    StructEndecBuilder.of1(
        mapEndecOf(Fragment.ENDEC, Fragment.ENDEC).fieldOf("entries", (fragment: MapFragment) => KtMap.getInstance().fromJsMap(fragment.map)),
        (map) => new MapFragment(map.asJsReadonlyMapView())
    )
);

File: src/jsMain/kotlin/me/maplesyrum/tricksterstudio/spell/fragment/MapFragment.kt
package me.maplesyrum.tricksterstudio.spell.fragment

import io.kvision.utils.obj
import me.maplesyrum.tricksterstudio.external.pixi.HTMLText
import tree.maple.kendec.StructEndec
import tree.maple.kendec.impl.StructEndecBuilder
import tree.maple.kendec.mapEndecOf


class MapFragment(val map: Map<Fragment, Fragment>) : Fragment() {
    override fun type(): FragmentType<*> {
        return FragmentType.MAP
    }

    override fun asFormattedText(): HTMLText {
        val out = buildString {
            append("{")
            val iterator = map.entries.iterator()
            while (iterator.hasNext()) {
                val (key, value) = iterator.next()
                append(key.asFormattedText().text)
                append(": ")
                append(value.asFormattedText().text)
                if (iterator.hasNext()) {
                    append(", ")
                }
            }
            append("}")
        }
        return HTMLText(
           obj {
               text = "<span style=\"color: #" + type().color.map{ it.toInt() }.orElseGet { 0xaaaaaa }.toString(16) + "\">$out</span>"
           }
        )
    }

    override fun toString(): String {
        return buildString {
            append("{")
            val iterator = map.entries.iterator()
            while (iterator.hasNext()) {
                val (key, value) = iterator.next()
                append(key.toString())
                append(": ")
                append(value.toString())
                if (iterator.hasNext()) {
                    append(", ")
                }
            }
            append("}")
        }
    }

    companion object {
        val ENDEC: StructEndec<MapFragment> = StructEndecBuilder.of(
            mapEndecOf(Fragment.ENDEC, Fragment.ENDEC)
                .fieldOf("entries", MapFragment::map)
        ) { map: Map<Fragment, Fragment> -> MapFragment(map) }
    }
}
